// pg-formatter/dist/format.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _child_process = require("child_process");

var _path = require("path");

var _shellQuote = require("shell-quote");

const executablePath = (0, _path.resolve)(__dirname, 'pg-formatter/pg_format');
const defaultConfiguration = {
  anonymize: false,
  functionCase: 'unchanged',
  keywordCase: 'unchanged',
  noRcFile: false,
  spaces: 4,
  stripComments: false,
  tabs: false
};

const createConfiguration = (userConfiguration = defaultConfiguration) => {
  // $FlowFixMe
  return { ...defaultConfiguration,
    ...userConfiguration
  };
};

const functionCaseOptionValueMap = {
  capitalize: 3,
  lowercase: 1,
  unchanged: 0,
  uppercase: 2
};
const keywordCaseOptionValueMap = functionCaseOptionValueMap;

const createCommandLineArgs = configuration => {
  const args = [];

  if (configuration.anonymize) {
    args.push('--anonymize');
  }

  if (configuration.functionCase) {
    args.push('--function-case ' + functionCaseOptionValueMap[configuration.functionCase]);
  }

  if (configuration.keywordCase) {
    args.push('--keyword-case ' + keywordCaseOptionValueMap[configuration.keywordCase]);
  }

  if (configuration.noRcFile) {
    args.push('--no-rcfile');
  }

  if (configuration.placeholder) {
    args.push('--placeholder ' + (0, _shellQuote.quote)([configuration.placeholder]));
  }

  if (configuration.spaces) {
    args.push('--spaces ' + configuration.spaces);
  }

  if (configuration.stripComments) {
    args.push('--nocomment');
  }

  if (configuration.tabs) {
    args.push('--tabs');
  }

  if (configuration.commaBreak) {
    args.push('--comma-break');
  }

  return args.join(' ');
};

var _default = (sql, userConfiguration) => {
  const configuration = createConfiguration(userConfiguration);
  const args = createCommandLineArgs(configuration);
  const result = (0, _child_process.execSync)('perl ' + executablePath + ' ' + args, {
    encoding: 'utf8',
    input: sql
  });
  return result;
};

exports.default = _default;
