// pgsql-parser/main/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseFunction = exports.Deparser = exports.deparse = exports.parseAsync = exports.parse = void 0;
const pgsql_deparser_1 = require("pgsql-deparser");
Object.defineProperty(exports, "Deparser", { enumerable: true, get: function () { return pgsql_deparser_1.Deparser; } });
Object.defineProperty(exports, "deparse", { enumerable: true, get: function () { return pgsql_deparser_1.deparse; } });
const libpg_query_1 = require("libpg-query");
Object.defineProperty(exports, "parseFunction", { enumerable: true, get: function () { return libpg_query_1.parsePlPgSQLSync; } });
function mapStmt({ stmt, stmt_len, stmt_location }) {
    return {
        RawStmt: {
            stmt,
            stmt_len,
            stmt_location: stmt_location || 0
        }
    };
}
const parse = (sql) => {
    if (!sql)
        throw new Error('no SQL provided to parser');
    const result = (0, libpg_query_1.parseQuerySync)(sql);
    return result.stmts.map(mapStmt);
};
exports.parse = parse;
const parseAsync = async (sql) => {
    if (!sql)
        throw new Error('no SQL provided to parser');
    const result = await (0, libpg_query_1.parseQuery)(sql);
    return result.stmts.map(mapStmt);
};
exports.parseAsync = parseAsync;
