// @pgsql/types/main/enums/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetOpStrategy = exports.SetOpCmd = exports.AggSplit = exports.AggStrategy = exports.JoinType = exports.CmdType = exports.BoolTestType = exports.NullTestType = exports.XmlOptionType = exports.XmlExprOp = exports.SQLValueFunctionOp = exports.MinMaxOp = exports.RowCompareType = exports.SubLinkType = exports.BoolExprType = exports.CoercionForm = exports.CoercionContext = exports.ParamKind = exports.OnCommitAction = exports.AlterSubscriptionType = exports.AlterTSConfigType = exports.ReindexObjectType = exports.DiscardMode = exports.ClusterOption = exports.ViewCheckOption = exports.TransactionStmtKind = exports.FunctionParameterMode = exports.FetchDirection = exports.RoleStmtType = exports.ImportForeignSchemaType = exports.ConstrType = exports.VariableSetKind = exports.GrantTargetType = exports.AlterTableType = exports.DropBehavior = exports.ObjectType = exports.SetOperation = exports.CTEMaterialize = exports.GroupingSetKind = exports.WCOKind = exports.RTEKind = exports.PartitionRangeDatumKind = exports.DefElemAction = exports.TableLikeOption = exports.RoleSpecType = exports.A_Expr_Kind = exports.SortByNulls = exports.SortByDir = exports.QuerySource = exports.OverridingKind = void 0;
exports.Token = exports.KeywordKind = exports.LockTupleMode = exports.LockWaitPolicy = exports.LockClauseStrength = exports.LimitOption = exports.OnConflictAction = void 0;
var OverridingKind;
(function (OverridingKind) {
    OverridingKind[OverridingKind["OVERRIDING_NOT_SET"] = 0] = "OVERRIDING_NOT_SET";
    OverridingKind[OverridingKind["OVERRIDING_USER_VALUE"] = 1] = "OVERRIDING_USER_VALUE";
    OverridingKind[OverridingKind["OVERRIDING_SYSTEM_VALUE"] = 2] = "OVERRIDING_SYSTEM_VALUE";
})(OverridingKind || (exports.OverridingKind = OverridingKind = {}));
var QuerySource;
(function (QuerySource) {
    QuerySource[QuerySource["QSRC_ORIGINAL"] = 0] = "QSRC_ORIGINAL";
    QuerySource[QuerySource["QSRC_PARSER"] = 1] = "QSRC_PARSER";
    QuerySource[QuerySource["QSRC_INSTEAD_RULE"] = 2] = "QSRC_INSTEAD_RULE";
    QuerySource[QuerySource["QSRC_QUAL_INSTEAD_RULE"] = 3] = "QSRC_QUAL_INSTEAD_RULE";
    QuerySource[QuerySource["QSRC_NON_INSTEAD_RULE"] = 4] = "QSRC_NON_INSTEAD_RULE";
})(QuerySource || (exports.QuerySource = QuerySource = {}));
var SortByDir;
(function (SortByDir) {
    SortByDir[SortByDir["SORTBY_DEFAULT"] = 0] = "SORTBY_DEFAULT";
    SortByDir[SortByDir["SORTBY_ASC"] = 1] = "SORTBY_ASC";
    SortByDir[SortByDir["SORTBY_DESC"] = 2] = "SORTBY_DESC";
    SortByDir[SortByDir["SORTBY_USING"] = 3] = "SORTBY_USING";
})(SortByDir || (exports.SortByDir = SortByDir = {}));
var SortByNulls;
(function (SortByNulls) {
    SortByNulls[SortByNulls["SORTBY_NULLS_DEFAULT"] = 0] = "SORTBY_NULLS_DEFAULT";
    SortByNulls[SortByNulls["SORTBY_NULLS_FIRST"] = 1] = "SORTBY_NULLS_FIRST";
    SortByNulls[SortByNulls["SORTBY_NULLS_LAST"] = 2] = "SORTBY_NULLS_LAST";
})(SortByNulls || (exports.SortByNulls = SortByNulls = {}));
var A_Expr_Kind;
(function (A_Expr_Kind) {
    A_Expr_Kind[A_Expr_Kind["AEXPR_OP"] = 0] = "AEXPR_OP";
    A_Expr_Kind[A_Expr_Kind["AEXPR_OP_ANY"] = 1] = "AEXPR_OP_ANY";
    A_Expr_Kind[A_Expr_Kind["AEXPR_OP_ALL"] = 2] = "AEXPR_OP_ALL";
    A_Expr_Kind[A_Expr_Kind["AEXPR_DISTINCT"] = 3] = "AEXPR_DISTINCT";
    A_Expr_Kind[A_Expr_Kind["AEXPR_NOT_DISTINCT"] = 4] = "AEXPR_NOT_DISTINCT";
    A_Expr_Kind[A_Expr_Kind["AEXPR_NULLIF"] = 5] = "AEXPR_NULLIF";
    A_Expr_Kind[A_Expr_Kind["AEXPR_OF"] = 6] = "AEXPR_OF";
    A_Expr_Kind[A_Expr_Kind["AEXPR_IN"] = 7] = "AEXPR_IN";
    A_Expr_Kind[A_Expr_Kind["AEXPR_LIKE"] = 8] = "AEXPR_LIKE";
    A_Expr_Kind[A_Expr_Kind["AEXPR_ILIKE"] = 9] = "AEXPR_ILIKE";
    A_Expr_Kind[A_Expr_Kind["AEXPR_SIMILAR"] = 10] = "AEXPR_SIMILAR";
    A_Expr_Kind[A_Expr_Kind["AEXPR_BETWEEN"] = 11] = "AEXPR_BETWEEN";
    A_Expr_Kind[A_Expr_Kind["AEXPR_NOT_BETWEEN"] = 12] = "AEXPR_NOT_BETWEEN";
    A_Expr_Kind[A_Expr_Kind["AEXPR_BETWEEN_SYM"] = 13] = "AEXPR_BETWEEN_SYM";
    A_Expr_Kind[A_Expr_Kind["AEXPR_NOT_BETWEEN_SYM"] = 14] = "AEXPR_NOT_BETWEEN_SYM";
    A_Expr_Kind[A_Expr_Kind["AEXPR_PAREN"] = 15] = "AEXPR_PAREN";
})(A_Expr_Kind || (exports.A_Expr_Kind = A_Expr_Kind = {}));
var RoleSpecType;
(function (RoleSpecType) {
    RoleSpecType[RoleSpecType["ROLESPEC_CSTRING"] = 0] = "ROLESPEC_CSTRING";
    RoleSpecType[RoleSpecType["ROLESPEC_CURRENT_USER"] = 1] = "ROLESPEC_CURRENT_USER";
    RoleSpecType[RoleSpecType["ROLESPEC_SESSION_USER"] = 2] = "ROLESPEC_SESSION_USER";
    RoleSpecType[RoleSpecType["ROLESPEC_PUBLIC"] = 3] = "ROLESPEC_PUBLIC";
})(RoleSpecType || (exports.RoleSpecType = RoleSpecType = {}));
var TableLikeOption;
(function (TableLikeOption) {
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_COMMENTS"] = 0] = "CREATE_TABLE_LIKE_COMMENTS";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_CONSTRAINTS"] = 1] = "CREATE_TABLE_LIKE_CONSTRAINTS";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_DEFAULTS"] = 2] = "CREATE_TABLE_LIKE_DEFAULTS";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_GENERATED"] = 3] = "CREATE_TABLE_LIKE_GENERATED";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_IDENTITY"] = 4] = "CREATE_TABLE_LIKE_IDENTITY";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_INDEXES"] = 5] = "CREATE_TABLE_LIKE_INDEXES";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_STATISTICS"] = 6] = "CREATE_TABLE_LIKE_STATISTICS";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_STORAGE"] = 7] = "CREATE_TABLE_LIKE_STORAGE";
    TableLikeOption[TableLikeOption["CREATE_TABLE_LIKE_ALL"] = 8] = "CREATE_TABLE_LIKE_ALL";
})(TableLikeOption || (exports.TableLikeOption = TableLikeOption = {}));
var DefElemAction;
(function (DefElemAction) {
    DefElemAction[DefElemAction["DEFELEM_UNSPEC"] = 0] = "DEFELEM_UNSPEC";
    DefElemAction[DefElemAction["DEFELEM_SET"] = 1] = "DEFELEM_SET";
    DefElemAction[DefElemAction["DEFELEM_ADD"] = 2] = "DEFELEM_ADD";
    DefElemAction[DefElemAction["DEFELEM_DROP"] = 3] = "DEFELEM_DROP";
})(DefElemAction || (exports.DefElemAction = DefElemAction = {}));
var PartitionRangeDatumKind;
(function (PartitionRangeDatumKind) {
    PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_MINVALUE"] = 0] = "PARTITION_RANGE_DATUM_MINVALUE";
    PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_VALUE"] = 1] = "PARTITION_RANGE_DATUM_VALUE";
    PartitionRangeDatumKind[PartitionRangeDatumKind["PARTITION_RANGE_DATUM_MAXVALUE"] = 2] = "PARTITION_RANGE_DATUM_MAXVALUE";
})(PartitionRangeDatumKind || (exports.PartitionRangeDatumKind = PartitionRangeDatumKind = {}));
var RTEKind;
(function (RTEKind) {
    RTEKind[RTEKind["RTE_RELATION"] = 0] = "RTE_RELATION";
    RTEKind[RTEKind["RTE_SUBQUERY"] = 1] = "RTE_SUBQUERY";
    RTEKind[RTEKind["RTE_JOIN"] = 2] = "RTE_JOIN";
    RTEKind[RTEKind["RTE_FUNCTION"] = 3] = "RTE_FUNCTION";
    RTEKind[RTEKind["RTE_TABLEFUNC"] = 4] = "RTE_TABLEFUNC";
    RTEKind[RTEKind["RTE_VALUES"] = 5] = "RTE_VALUES";
    RTEKind[RTEKind["RTE_CTE"] = 6] = "RTE_CTE";
    RTEKind[RTEKind["RTE_NAMEDTUPLESTORE"] = 7] = "RTE_NAMEDTUPLESTORE";
    RTEKind[RTEKind["RTE_RESULT"] = 8] = "RTE_RESULT";
})(RTEKind || (exports.RTEKind = RTEKind = {}));
var WCOKind;
(function (WCOKind) {
    WCOKind[WCOKind["WCO_VIEW_CHECK"] = 0] = "WCO_VIEW_CHECK";
    WCOKind[WCOKind["WCO_RLS_INSERT_CHECK"] = 1] = "WCO_RLS_INSERT_CHECK";
    WCOKind[WCOKind["WCO_RLS_UPDATE_CHECK"] = 2] = "WCO_RLS_UPDATE_CHECK";
    WCOKind[WCOKind["WCO_RLS_CONFLICT_CHECK"] = 3] = "WCO_RLS_CONFLICT_CHECK";
})(WCOKind || (exports.WCOKind = WCOKind = {}));
var GroupingSetKind;
(function (GroupingSetKind) {
    GroupingSetKind[GroupingSetKind["GROUPING_SET_EMPTY"] = 0] = "GROUPING_SET_EMPTY";
    GroupingSetKind[GroupingSetKind["GROUPING_SET_SIMPLE"] = 1] = "GROUPING_SET_SIMPLE";
    GroupingSetKind[GroupingSetKind["GROUPING_SET_ROLLUP"] = 2] = "GROUPING_SET_ROLLUP";
    GroupingSetKind[GroupingSetKind["GROUPING_SET_CUBE"] = 3] = "GROUPING_SET_CUBE";
    GroupingSetKind[GroupingSetKind["GROUPING_SET_SETS"] = 4] = "GROUPING_SET_SETS";
})(GroupingSetKind || (exports.GroupingSetKind = GroupingSetKind = {}));
var CTEMaterialize;
(function (CTEMaterialize) {
    CTEMaterialize[CTEMaterialize["CTEMaterializeDefault"] = 0] = "CTEMaterializeDefault";
    CTEMaterialize[CTEMaterialize["CTEMaterializeAlways"] = 1] = "CTEMaterializeAlways";
    CTEMaterialize[CTEMaterialize["CTEMaterializeNever"] = 2] = "CTEMaterializeNever";
})(CTEMaterialize || (exports.CTEMaterialize = CTEMaterialize = {}));
var SetOperation;
(function (SetOperation) {
    SetOperation[SetOperation["SETOP_NONE"] = 0] = "SETOP_NONE";
    SetOperation[SetOperation["SETOP_UNION"] = 1] = "SETOP_UNION";
    SetOperation[SetOperation["SETOP_INTERSECT"] = 2] = "SETOP_INTERSECT";
    SetOperation[SetOperation["SETOP_EXCEPT"] = 3] = "SETOP_EXCEPT";
})(SetOperation || (exports.SetOperation = SetOperation = {}));
var ObjectType;
(function (ObjectType) {
    ObjectType[ObjectType["OBJECT_ACCESS_METHOD"] = 0] = "OBJECT_ACCESS_METHOD";
    ObjectType[ObjectType["OBJECT_AGGREGATE"] = 1] = "OBJECT_AGGREGATE";
    ObjectType[ObjectType["OBJECT_AMOP"] = 2] = "OBJECT_AMOP";
    ObjectType[ObjectType["OBJECT_AMPROC"] = 3] = "OBJECT_AMPROC";
    ObjectType[ObjectType["OBJECT_ATTRIBUTE"] = 4] = "OBJECT_ATTRIBUTE";
    ObjectType[ObjectType["OBJECT_CAST"] = 5] = "OBJECT_CAST";
    ObjectType[ObjectType["OBJECT_COLUMN"] = 6] = "OBJECT_COLUMN";
    ObjectType[ObjectType["OBJECT_COLLATION"] = 7] = "OBJECT_COLLATION";
    ObjectType[ObjectType["OBJECT_CONVERSION"] = 8] = "OBJECT_CONVERSION";
    ObjectType[ObjectType["OBJECT_DATABASE"] = 9] = "OBJECT_DATABASE";
    ObjectType[ObjectType["OBJECT_DEFAULT"] = 10] = "OBJECT_DEFAULT";
    ObjectType[ObjectType["OBJECT_DEFACL"] = 11] = "OBJECT_DEFACL";
    ObjectType[ObjectType["OBJECT_DOMAIN"] = 12] = "OBJECT_DOMAIN";
    ObjectType[ObjectType["OBJECT_DOMCONSTRAINT"] = 13] = "OBJECT_DOMCONSTRAINT";
    ObjectType[ObjectType["OBJECT_EVENT_TRIGGER"] = 14] = "OBJECT_EVENT_TRIGGER";
    ObjectType[ObjectType["OBJECT_EXTENSION"] = 15] = "OBJECT_EXTENSION";
    ObjectType[ObjectType["OBJECT_FDW"] = 16] = "OBJECT_FDW";
    ObjectType[ObjectType["OBJECT_FOREIGN_SERVER"] = 17] = "OBJECT_FOREIGN_SERVER";
    ObjectType[ObjectType["OBJECT_FOREIGN_TABLE"] = 18] = "OBJECT_FOREIGN_TABLE";
    ObjectType[ObjectType["OBJECT_FUNCTION"] = 19] = "OBJECT_FUNCTION";
    ObjectType[ObjectType["OBJECT_INDEX"] = 20] = "OBJECT_INDEX";
    ObjectType[ObjectType["OBJECT_LANGUAGE"] = 21] = "OBJECT_LANGUAGE";
    ObjectType[ObjectType["OBJECT_LARGEOBJECT"] = 22] = "OBJECT_LARGEOBJECT";
    ObjectType[ObjectType["OBJECT_MATVIEW"] = 23] = "OBJECT_MATVIEW";
    ObjectType[ObjectType["OBJECT_OPCLASS"] = 24] = "OBJECT_OPCLASS";
    ObjectType[ObjectType["OBJECT_OPERATOR"] = 25] = "OBJECT_OPERATOR";
    ObjectType[ObjectType["OBJECT_OPFAMILY"] = 26] = "OBJECT_OPFAMILY";
    ObjectType[ObjectType["OBJECT_POLICY"] = 27] = "OBJECT_POLICY";
    ObjectType[ObjectType["OBJECT_PROCEDURE"] = 28] = "OBJECT_PROCEDURE";
    ObjectType[ObjectType["OBJECT_PUBLICATION"] = 29] = "OBJECT_PUBLICATION";
    ObjectType[ObjectType["OBJECT_PUBLICATION_REL"] = 30] = "OBJECT_PUBLICATION_REL";
    ObjectType[ObjectType["OBJECT_ROLE"] = 31] = "OBJECT_ROLE";
    ObjectType[ObjectType["OBJECT_ROUTINE"] = 32] = "OBJECT_ROUTINE";
    ObjectType[ObjectType["OBJECT_RULE"] = 33] = "OBJECT_RULE";
    ObjectType[ObjectType["OBJECT_SCHEMA"] = 34] = "OBJECT_SCHEMA";
    ObjectType[ObjectType["OBJECT_SEQUENCE"] = 35] = "OBJECT_SEQUENCE";
    ObjectType[ObjectType["OBJECT_SUBSCRIPTION"] = 36] = "OBJECT_SUBSCRIPTION";
    ObjectType[ObjectType["OBJECT_STATISTIC_EXT"] = 37] = "OBJECT_STATISTIC_EXT";
    ObjectType[ObjectType["OBJECT_TABCONSTRAINT"] = 38] = "OBJECT_TABCONSTRAINT";
    ObjectType[ObjectType["OBJECT_TABLE"] = 39] = "OBJECT_TABLE";
    ObjectType[ObjectType["OBJECT_TABLESPACE"] = 40] = "OBJECT_TABLESPACE";
    ObjectType[ObjectType["OBJECT_TRANSFORM"] = 41] = "OBJECT_TRANSFORM";
    ObjectType[ObjectType["OBJECT_TRIGGER"] = 42] = "OBJECT_TRIGGER";
    ObjectType[ObjectType["OBJECT_TSCONFIGURATION"] = 43] = "OBJECT_TSCONFIGURATION";
    ObjectType[ObjectType["OBJECT_TSDICTIONARY"] = 44] = "OBJECT_TSDICTIONARY";
    ObjectType[ObjectType["OBJECT_TSPARSER"] = 45] = "OBJECT_TSPARSER";
    ObjectType[ObjectType["OBJECT_TSTEMPLATE"] = 46] = "OBJECT_TSTEMPLATE";
    ObjectType[ObjectType["OBJECT_TYPE"] = 47] = "OBJECT_TYPE";
    ObjectType[ObjectType["OBJECT_USER_MAPPING"] = 48] = "OBJECT_USER_MAPPING";
    ObjectType[ObjectType["OBJECT_VIEW"] = 49] = "OBJECT_VIEW";
})(ObjectType || (exports.ObjectType = ObjectType = {}));
var DropBehavior;
(function (DropBehavior) {
    DropBehavior[DropBehavior["DROP_RESTRICT"] = 0] = "DROP_RESTRICT";
    DropBehavior[DropBehavior["DROP_CASCADE"] = 1] = "DROP_CASCADE";
})(DropBehavior || (exports.DropBehavior = DropBehavior = {}));
var AlterTableType;
(function (AlterTableType) {
    AlterTableType[AlterTableType["AT_AddColumn"] = 0] = "AT_AddColumn";
    AlterTableType[AlterTableType["AT_AddColumnRecurse"] = 1] = "AT_AddColumnRecurse";
    AlterTableType[AlterTableType["AT_AddColumnToView"] = 2] = "AT_AddColumnToView";
    AlterTableType[AlterTableType["AT_ColumnDefault"] = 3] = "AT_ColumnDefault";
    AlterTableType[AlterTableType["AT_CookedColumnDefault"] = 4] = "AT_CookedColumnDefault";
    AlterTableType[AlterTableType["AT_DropNotNull"] = 5] = "AT_DropNotNull";
    AlterTableType[AlterTableType["AT_SetNotNull"] = 6] = "AT_SetNotNull";
    AlterTableType[AlterTableType["AT_DropExpression"] = 7] = "AT_DropExpression";
    AlterTableType[AlterTableType["AT_CheckNotNull"] = 8] = "AT_CheckNotNull";
    AlterTableType[AlterTableType["AT_SetStatistics"] = 9] = "AT_SetStatistics";
    AlterTableType[AlterTableType["AT_SetOptions"] = 10] = "AT_SetOptions";
    AlterTableType[AlterTableType["AT_ResetOptions"] = 11] = "AT_ResetOptions";
    AlterTableType[AlterTableType["AT_SetStorage"] = 12] = "AT_SetStorage";
    AlterTableType[AlterTableType["AT_DropColumn"] = 13] = "AT_DropColumn";
    AlterTableType[AlterTableType["AT_DropColumnRecurse"] = 14] = "AT_DropColumnRecurse";
    AlterTableType[AlterTableType["AT_AddIndex"] = 15] = "AT_AddIndex";
    AlterTableType[AlterTableType["AT_ReAddIndex"] = 16] = "AT_ReAddIndex";
    AlterTableType[AlterTableType["AT_AddConstraint"] = 17] = "AT_AddConstraint";
    AlterTableType[AlterTableType["AT_AddConstraintRecurse"] = 18] = "AT_AddConstraintRecurse";
    AlterTableType[AlterTableType["AT_ReAddConstraint"] = 19] = "AT_ReAddConstraint";
    AlterTableType[AlterTableType["AT_ReAddDomainConstraint"] = 20] = "AT_ReAddDomainConstraint";
    AlterTableType[AlterTableType["AT_AlterConstraint"] = 21] = "AT_AlterConstraint";
    AlterTableType[AlterTableType["AT_ValidateConstraint"] = 22] = "AT_ValidateConstraint";
    AlterTableType[AlterTableType["AT_ValidateConstraintRecurse"] = 23] = "AT_ValidateConstraintRecurse";
    AlterTableType[AlterTableType["AT_AddIndexConstraint"] = 24] = "AT_AddIndexConstraint";
    AlterTableType[AlterTableType["AT_DropConstraint"] = 25] = "AT_DropConstraint";
    AlterTableType[AlterTableType["AT_DropConstraintRecurse"] = 26] = "AT_DropConstraintRecurse";
    AlterTableType[AlterTableType["AT_ReAddComment"] = 27] = "AT_ReAddComment";
    AlterTableType[AlterTableType["AT_AlterColumnType"] = 28] = "AT_AlterColumnType";
    AlterTableType[AlterTableType["AT_AlterColumnGenericOptions"] = 29] = "AT_AlterColumnGenericOptions";
    AlterTableType[AlterTableType["AT_ChangeOwner"] = 30] = "AT_ChangeOwner";
    AlterTableType[AlterTableType["AT_ClusterOn"] = 31] = "AT_ClusterOn";
    AlterTableType[AlterTableType["AT_DropCluster"] = 32] = "AT_DropCluster";
    AlterTableType[AlterTableType["AT_SetLogged"] = 33] = "AT_SetLogged";
    AlterTableType[AlterTableType["AT_SetUnLogged"] = 34] = "AT_SetUnLogged";
    AlterTableType[AlterTableType["AT_DropOids"] = 35] = "AT_DropOids";
    AlterTableType[AlterTableType["AT_SetTableSpace"] = 36] = "AT_SetTableSpace";
    AlterTableType[AlterTableType["AT_SetRelOptions"] = 37] = "AT_SetRelOptions";
    AlterTableType[AlterTableType["AT_ResetRelOptions"] = 38] = "AT_ResetRelOptions";
    AlterTableType[AlterTableType["AT_ReplaceRelOptions"] = 39] = "AT_ReplaceRelOptions";
    AlterTableType[AlterTableType["AT_EnableTrig"] = 40] = "AT_EnableTrig";
    AlterTableType[AlterTableType["AT_EnableAlwaysTrig"] = 41] = "AT_EnableAlwaysTrig";
    AlterTableType[AlterTableType["AT_EnableReplicaTrig"] = 42] = "AT_EnableReplicaTrig";
    AlterTableType[AlterTableType["AT_DisableTrig"] = 43] = "AT_DisableTrig";
    AlterTableType[AlterTableType["AT_EnableTrigAll"] = 44] = "AT_EnableTrigAll";
    AlterTableType[AlterTableType["AT_DisableTrigAll"] = 45] = "AT_DisableTrigAll";
    AlterTableType[AlterTableType["AT_EnableTrigUser"] = 46] = "AT_EnableTrigUser";
    AlterTableType[AlterTableType["AT_DisableTrigUser"] = 47] = "AT_DisableTrigUser";
    AlterTableType[AlterTableType["AT_EnableRule"] = 48] = "AT_EnableRule";
    AlterTableType[AlterTableType["AT_EnableAlwaysRule"] = 49] = "AT_EnableAlwaysRule";
    AlterTableType[AlterTableType["AT_EnableReplicaRule"] = 50] = "AT_EnableReplicaRule";
    AlterTableType[AlterTableType["AT_DisableRule"] = 51] = "AT_DisableRule";
    AlterTableType[AlterTableType["AT_AddInherit"] = 52] = "AT_AddInherit";
    AlterTableType[AlterTableType["AT_DropInherit"] = 53] = "AT_DropInherit";
    AlterTableType[AlterTableType["AT_AddOf"] = 54] = "AT_AddOf";
    AlterTableType[AlterTableType["AT_DropOf"] = 55] = "AT_DropOf";
    AlterTableType[AlterTableType["AT_ReplicaIdentity"] = 56] = "AT_ReplicaIdentity";
    AlterTableType[AlterTableType["AT_EnableRowSecurity"] = 57] = "AT_EnableRowSecurity";
    AlterTableType[AlterTableType["AT_DisableRowSecurity"] = 58] = "AT_DisableRowSecurity";
    AlterTableType[AlterTableType["AT_ForceRowSecurity"] = 59] = "AT_ForceRowSecurity";
    AlterTableType[AlterTableType["AT_NoForceRowSecurity"] = 60] = "AT_NoForceRowSecurity";
    AlterTableType[AlterTableType["AT_GenericOptions"] = 61] = "AT_GenericOptions";
    AlterTableType[AlterTableType["AT_AttachPartition"] = 62] = "AT_AttachPartition";
    AlterTableType[AlterTableType["AT_DetachPartition"] = 63] = "AT_DetachPartition";
    AlterTableType[AlterTableType["AT_AddIdentity"] = 64] = "AT_AddIdentity";
    AlterTableType[AlterTableType["AT_SetIdentity"] = 65] = "AT_SetIdentity";
    AlterTableType[AlterTableType["AT_DropIdentity"] = 66] = "AT_DropIdentity";
})(AlterTableType || (exports.AlterTableType = AlterTableType = {}));
var GrantTargetType;
(function (GrantTargetType) {
    GrantTargetType[GrantTargetType["ACL_TARGET_OBJECT"] = 0] = "ACL_TARGET_OBJECT";
    GrantTargetType[GrantTargetType["ACL_TARGET_ALL_IN_SCHEMA"] = 1] = "ACL_TARGET_ALL_IN_SCHEMA";
    GrantTargetType[GrantTargetType["ACL_TARGET_DEFAULTS"] = 2] = "ACL_TARGET_DEFAULTS";
})(GrantTargetType || (exports.GrantTargetType = GrantTargetType = {}));
var VariableSetKind;
(function (VariableSetKind) {
    VariableSetKind[VariableSetKind["VAR_SET_VALUE"] = 0] = "VAR_SET_VALUE";
    VariableSetKind[VariableSetKind["VAR_SET_DEFAULT"] = 1] = "VAR_SET_DEFAULT";
    VariableSetKind[VariableSetKind["VAR_SET_CURRENT"] = 2] = "VAR_SET_CURRENT";
    VariableSetKind[VariableSetKind["VAR_SET_MULTI"] = 3] = "VAR_SET_MULTI";
    VariableSetKind[VariableSetKind["VAR_RESET"] = 4] = "VAR_RESET";
    VariableSetKind[VariableSetKind["VAR_RESET_ALL"] = 5] = "VAR_RESET_ALL";
})(VariableSetKind || (exports.VariableSetKind = VariableSetKind = {}));
var ConstrType;
(function (ConstrType) {
    ConstrType[ConstrType["CONSTR_NULL"] = 0] = "CONSTR_NULL";
    ConstrType[ConstrType["CONSTR_NOTNULL"] = 1] = "CONSTR_NOTNULL";
    ConstrType[ConstrType["CONSTR_DEFAULT"] = 2] = "CONSTR_DEFAULT";
    ConstrType[ConstrType["CONSTR_IDENTITY"] = 3] = "CONSTR_IDENTITY";
    ConstrType[ConstrType["CONSTR_GENERATED"] = 4] = "CONSTR_GENERATED";
    ConstrType[ConstrType["CONSTR_CHECK"] = 5] = "CONSTR_CHECK";
    ConstrType[ConstrType["CONSTR_PRIMARY"] = 6] = "CONSTR_PRIMARY";
    ConstrType[ConstrType["CONSTR_UNIQUE"] = 7] = "CONSTR_UNIQUE";
    ConstrType[ConstrType["CONSTR_EXCLUSION"] = 8] = "CONSTR_EXCLUSION";
    ConstrType[ConstrType["CONSTR_FOREIGN"] = 9] = "CONSTR_FOREIGN";
    ConstrType[ConstrType["CONSTR_ATTR_DEFERRABLE"] = 10] = "CONSTR_ATTR_DEFERRABLE";
    ConstrType[ConstrType["CONSTR_ATTR_NOT_DEFERRABLE"] = 11] = "CONSTR_ATTR_NOT_DEFERRABLE";
    ConstrType[ConstrType["CONSTR_ATTR_DEFERRED"] = 12] = "CONSTR_ATTR_DEFERRED";
    ConstrType[ConstrType["CONSTR_ATTR_IMMEDIATE"] = 13] = "CONSTR_ATTR_IMMEDIATE";
})(ConstrType || (exports.ConstrType = ConstrType = {}));
var ImportForeignSchemaType;
(function (ImportForeignSchemaType) {
    ImportForeignSchemaType[ImportForeignSchemaType["FDW_IMPORT_SCHEMA_ALL"] = 0] = "FDW_IMPORT_SCHEMA_ALL";
    ImportForeignSchemaType[ImportForeignSchemaType["FDW_IMPORT_SCHEMA_LIMIT_TO"] = 1] = "FDW_IMPORT_SCHEMA_LIMIT_TO";
    ImportForeignSchemaType[ImportForeignSchemaType["FDW_IMPORT_SCHEMA_EXCEPT"] = 2] = "FDW_IMPORT_SCHEMA_EXCEPT";
})(ImportForeignSchemaType || (exports.ImportForeignSchemaType = ImportForeignSchemaType = {}));
var RoleStmtType;
(function (RoleStmtType) {
    RoleStmtType[RoleStmtType["ROLESTMT_ROLE"] = 0] = "ROLESTMT_ROLE";
    RoleStmtType[RoleStmtType["ROLESTMT_USER"] = 1] = "ROLESTMT_USER";
    RoleStmtType[RoleStmtType["ROLESTMT_GROUP"] = 2] = "ROLESTMT_GROUP";
})(RoleStmtType || (exports.RoleStmtType = RoleStmtType = {}));
var FetchDirection;
(function (FetchDirection) {
    FetchDirection[FetchDirection["FETCH_FORWARD"] = 0] = "FETCH_FORWARD";
    FetchDirection[FetchDirection["FETCH_BACKWARD"] = 1] = "FETCH_BACKWARD";
    FetchDirection[FetchDirection["FETCH_ABSOLUTE"] = 2] = "FETCH_ABSOLUTE";
    FetchDirection[FetchDirection["FETCH_RELATIVE"] = 3] = "FETCH_RELATIVE";
})(FetchDirection || (exports.FetchDirection = FetchDirection = {}));
var FunctionParameterMode;
(function (FunctionParameterMode) {
    FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_IN"] = 0] = "FUNC_PARAM_IN";
    FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_OUT"] = 1] = "FUNC_PARAM_OUT";
    FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_INOUT"] = 2] = "FUNC_PARAM_INOUT";
    FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_VARIADIC"] = 3] = "FUNC_PARAM_VARIADIC";
    FunctionParameterMode[FunctionParameterMode["FUNC_PARAM_TABLE"] = 4] = "FUNC_PARAM_TABLE";
})(FunctionParameterMode || (exports.FunctionParameterMode = FunctionParameterMode = {}));
var TransactionStmtKind;
(function (TransactionStmtKind) {
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_BEGIN"] = 0] = "TRANS_STMT_BEGIN";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_START"] = 1] = "TRANS_STMT_START";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_COMMIT"] = 2] = "TRANS_STMT_COMMIT";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_ROLLBACK"] = 3] = "TRANS_STMT_ROLLBACK";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_SAVEPOINT"] = 4] = "TRANS_STMT_SAVEPOINT";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_RELEASE"] = 5] = "TRANS_STMT_RELEASE";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_ROLLBACK_TO"] = 6] = "TRANS_STMT_ROLLBACK_TO";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_PREPARE"] = 7] = "TRANS_STMT_PREPARE";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_COMMIT_PREPARED"] = 8] = "TRANS_STMT_COMMIT_PREPARED";
    TransactionStmtKind[TransactionStmtKind["TRANS_STMT_ROLLBACK_PREPARED"] = 9] = "TRANS_STMT_ROLLBACK_PREPARED";
})(TransactionStmtKind || (exports.TransactionStmtKind = TransactionStmtKind = {}));
var ViewCheckOption;
(function (ViewCheckOption) {
    ViewCheckOption[ViewCheckOption["NO_CHECK_OPTION"] = 0] = "NO_CHECK_OPTION";
    ViewCheckOption[ViewCheckOption["LOCAL_CHECK_OPTION"] = 1] = "LOCAL_CHECK_OPTION";
    ViewCheckOption[ViewCheckOption["CASCADED_CHECK_OPTION"] = 2] = "CASCADED_CHECK_OPTION";
})(ViewCheckOption || (exports.ViewCheckOption = ViewCheckOption = {}));
var ClusterOption;
(function (ClusterOption) {
    ClusterOption[ClusterOption["CLUOPT_RECHECK"] = 0] = "CLUOPT_RECHECK";
    ClusterOption[ClusterOption["CLUOPT_VERBOSE"] = 1] = "CLUOPT_VERBOSE";
})(ClusterOption || (exports.ClusterOption = ClusterOption = {}));
var DiscardMode;
(function (DiscardMode) {
    DiscardMode[DiscardMode["DISCARD_ALL"] = 0] = "DISCARD_ALL";
    DiscardMode[DiscardMode["DISCARD_PLANS"] = 1] = "DISCARD_PLANS";
    DiscardMode[DiscardMode["DISCARD_SEQUENCES"] = 2] = "DISCARD_SEQUENCES";
    DiscardMode[DiscardMode["DISCARD_TEMP"] = 3] = "DISCARD_TEMP";
})(DiscardMode || (exports.DiscardMode = DiscardMode = {}));
var ReindexObjectType;
(function (ReindexObjectType) {
    ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_INDEX"] = 0] = "REINDEX_OBJECT_INDEX";
    ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_TABLE"] = 1] = "REINDEX_OBJECT_TABLE";
    ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_SCHEMA"] = 2] = "REINDEX_OBJECT_SCHEMA";
    ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_SYSTEM"] = 3] = "REINDEX_OBJECT_SYSTEM";
    ReindexObjectType[ReindexObjectType["REINDEX_OBJECT_DATABASE"] = 4] = "REINDEX_OBJECT_DATABASE";
})(ReindexObjectType || (exports.ReindexObjectType = ReindexObjectType = {}));
var AlterTSConfigType;
(function (AlterTSConfigType) {
    AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_ADD_MAPPING"] = 0] = "ALTER_TSCONFIG_ADD_MAPPING";
    AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_ALTER_MAPPING_FOR_TOKEN"] = 1] = "ALTER_TSCONFIG_ALTER_MAPPING_FOR_TOKEN";
    AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_REPLACE_DICT"] = 2] = "ALTER_TSCONFIG_REPLACE_DICT";
    AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_REPLACE_DICT_FOR_TOKEN"] = 3] = "ALTER_TSCONFIG_REPLACE_DICT_FOR_TOKEN";
    AlterTSConfigType[AlterTSConfigType["ALTER_TSCONFIG_DROP_MAPPING"] = 4] = "ALTER_TSCONFIG_DROP_MAPPING";
})(AlterTSConfigType || (exports.AlterTSConfigType = AlterTSConfigType = {}));
var AlterSubscriptionType;
(function (AlterSubscriptionType) {
    AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_OPTIONS"] = 0] = "ALTER_SUBSCRIPTION_OPTIONS";
    AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_CONNECTION"] = 1] = "ALTER_SUBSCRIPTION_CONNECTION";
    AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_PUBLICATION"] = 2] = "ALTER_SUBSCRIPTION_PUBLICATION";
    AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_REFRESH"] = 3] = "ALTER_SUBSCRIPTION_REFRESH";
    AlterSubscriptionType[AlterSubscriptionType["ALTER_SUBSCRIPTION_ENABLED"] = 4] = "ALTER_SUBSCRIPTION_ENABLED";
})(AlterSubscriptionType || (exports.AlterSubscriptionType = AlterSubscriptionType = {}));
var OnCommitAction;
(function (OnCommitAction) {
    OnCommitAction[OnCommitAction["ONCOMMIT_NOOP"] = 0] = "ONCOMMIT_NOOP";
    OnCommitAction[OnCommitAction["ONCOMMIT_PRESERVE_ROWS"] = 1] = "ONCOMMIT_PRESERVE_ROWS";
    OnCommitAction[OnCommitAction["ONCOMMIT_DELETE_ROWS"] = 2] = "ONCOMMIT_DELETE_ROWS";
    OnCommitAction[OnCommitAction["ONCOMMIT_DROP"] = 3] = "ONCOMMIT_DROP";
})(OnCommitAction || (exports.OnCommitAction = OnCommitAction = {}));
var ParamKind;
(function (ParamKind) {
    ParamKind[ParamKind["PARAM_EXTERN"] = 0] = "PARAM_EXTERN";
    ParamKind[ParamKind["PARAM_EXEC"] = 1] = "PARAM_EXEC";
    ParamKind[ParamKind["PARAM_SUBLINK"] = 2] = "PARAM_SUBLINK";
    ParamKind[ParamKind["PARAM_MULTIEXPR"] = 3] = "PARAM_MULTIEXPR";
})(ParamKind || (exports.ParamKind = ParamKind = {}));
var CoercionContext;
(function (CoercionContext) {
    CoercionContext[CoercionContext["COERCION_IMPLICIT"] = 0] = "COERCION_IMPLICIT";
    CoercionContext[CoercionContext["COERCION_ASSIGNMENT"] = 1] = "COERCION_ASSIGNMENT";
    CoercionContext[CoercionContext["COERCION_EXPLICIT"] = 2] = "COERCION_EXPLICIT";
})(CoercionContext || (exports.CoercionContext = CoercionContext = {}));
var CoercionForm;
(function (CoercionForm) {
    CoercionForm[CoercionForm["COERCE_EXPLICIT_CALL"] = 0] = "COERCE_EXPLICIT_CALL";
    CoercionForm[CoercionForm["COERCE_EXPLICIT_CAST"] = 1] = "COERCE_EXPLICIT_CAST";
    CoercionForm[CoercionForm["COERCE_IMPLICIT_CAST"] = 2] = "COERCE_IMPLICIT_CAST";
})(CoercionForm || (exports.CoercionForm = CoercionForm = {}));
var BoolExprType;
(function (BoolExprType) {
    BoolExprType[BoolExprType["AND_EXPR"] = 0] = "AND_EXPR";
    BoolExprType[BoolExprType["OR_EXPR"] = 1] = "OR_EXPR";
    BoolExprType[BoolExprType["NOT_EXPR"] = 2] = "NOT_EXPR";
})(BoolExprType || (exports.BoolExprType = BoolExprType = {}));
var SubLinkType;
(function (SubLinkType) {
    SubLinkType[SubLinkType["EXISTS_SUBLINK"] = 0] = "EXISTS_SUBLINK";
    SubLinkType[SubLinkType["ALL_SUBLINK"] = 1] = "ALL_SUBLINK";
    SubLinkType[SubLinkType["ANY_SUBLINK"] = 2] = "ANY_SUBLINK";
    SubLinkType[SubLinkType["ROWCOMPARE_SUBLINK"] = 3] = "ROWCOMPARE_SUBLINK";
    SubLinkType[SubLinkType["EXPR_SUBLINK"] = 4] = "EXPR_SUBLINK";
    SubLinkType[SubLinkType["MULTIEXPR_SUBLINK"] = 5] = "MULTIEXPR_SUBLINK";
    SubLinkType[SubLinkType["ARRAY_SUBLINK"] = 6] = "ARRAY_SUBLINK";
    SubLinkType[SubLinkType["CTE_SUBLINK"] = 7] = "CTE_SUBLINK";
})(SubLinkType || (exports.SubLinkType = SubLinkType = {}));
var RowCompareType;
(function (RowCompareType) {
    RowCompareType[RowCompareType["ROWCOMPARE_LT"] = 0] = "ROWCOMPARE_LT";
    RowCompareType[RowCompareType["ROWCOMPARE_LE"] = 1] = "ROWCOMPARE_LE";
    RowCompareType[RowCompareType["ROWCOMPARE_EQ"] = 2] = "ROWCOMPARE_EQ";
    RowCompareType[RowCompareType["ROWCOMPARE_GE"] = 3] = "ROWCOMPARE_GE";
    RowCompareType[RowCompareType["ROWCOMPARE_GT"] = 4] = "ROWCOMPARE_GT";
    RowCompareType[RowCompareType["ROWCOMPARE_NE"] = 5] = "ROWCOMPARE_NE";
})(RowCompareType || (exports.RowCompareType = RowCompareType = {}));
var MinMaxOp;
(function (MinMaxOp) {
    MinMaxOp[MinMaxOp["IS_GREATEST"] = 0] = "IS_GREATEST";
    MinMaxOp[MinMaxOp["IS_LEAST"] = 1] = "IS_LEAST";
})(MinMaxOp || (exports.MinMaxOp = MinMaxOp = {}));
var SQLValueFunctionOp;
(function (SQLValueFunctionOp) {
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_DATE"] = 0] = "SVFOP_CURRENT_DATE";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIME"] = 1] = "SVFOP_CURRENT_TIME";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIME_N"] = 2] = "SVFOP_CURRENT_TIME_N";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIMESTAMP"] = 3] = "SVFOP_CURRENT_TIMESTAMP";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_TIMESTAMP_N"] = 4] = "SVFOP_CURRENT_TIMESTAMP_N";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIME"] = 5] = "SVFOP_LOCALTIME";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIME_N"] = 6] = "SVFOP_LOCALTIME_N";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIMESTAMP"] = 7] = "SVFOP_LOCALTIMESTAMP";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_LOCALTIMESTAMP_N"] = 8] = "SVFOP_LOCALTIMESTAMP_N";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_ROLE"] = 9] = "SVFOP_CURRENT_ROLE";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_USER"] = 10] = "SVFOP_CURRENT_USER";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_USER"] = 11] = "SVFOP_USER";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_SESSION_USER"] = 12] = "SVFOP_SESSION_USER";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_CATALOG"] = 13] = "SVFOP_CURRENT_CATALOG";
    SQLValueFunctionOp[SQLValueFunctionOp["SVFOP_CURRENT_SCHEMA"] = 14] = "SVFOP_CURRENT_SCHEMA";
})(SQLValueFunctionOp || (exports.SQLValueFunctionOp = SQLValueFunctionOp = {}));
var XmlExprOp;
(function (XmlExprOp) {
    XmlExprOp[XmlExprOp["IS_XMLCONCAT"] = 0] = "IS_XMLCONCAT";
    XmlExprOp[XmlExprOp["IS_XMLELEMENT"] = 1] = "IS_XMLELEMENT";
    XmlExprOp[XmlExprOp["IS_XMLFOREST"] = 2] = "IS_XMLFOREST";
    XmlExprOp[XmlExprOp["IS_XMLPARSE"] = 3] = "IS_XMLPARSE";
    XmlExprOp[XmlExprOp["IS_XMLPI"] = 4] = "IS_XMLPI";
    XmlExprOp[XmlExprOp["IS_XMLROOT"] = 5] = "IS_XMLROOT";
    XmlExprOp[XmlExprOp["IS_XMLSERIALIZE"] = 6] = "IS_XMLSERIALIZE";
    XmlExprOp[XmlExprOp["IS_DOCUMENT"] = 7] = "IS_DOCUMENT";
})(XmlExprOp || (exports.XmlExprOp = XmlExprOp = {}));
var XmlOptionType;
(function (XmlOptionType) {
    XmlOptionType[XmlOptionType["XMLOPTION_DOCUMENT"] = 0] = "XMLOPTION_DOCUMENT";
    XmlOptionType[XmlOptionType["XMLOPTION_CONTENT"] = 1] = "XMLOPTION_CONTENT";
})(XmlOptionType || (exports.XmlOptionType = XmlOptionType = {}));
var NullTestType;
(function (NullTestType) {
    NullTestType[NullTestType["IS_NULL"] = 0] = "IS_NULL";
    NullTestType[NullTestType["IS_NOT_NULL"] = 1] = "IS_NOT_NULL";
})(NullTestType || (exports.NullTestType = NullTestType = {}));
var BoolTestType;
(function (BoolTestType) {
    BoolTestType[BoolTestType["IS_TRUE"] = 0] = "IS_TRUE";
    BoolTestType[BoolTestType["IS_NOT_TRUE"] = 1] = "IS_NOT_TRUE";
    BoolTestType[BoolTestType["IS_FALSE"] = 2] = "IS_FALSE";
    BoolTestType[BoolTestType["IS_NOT_FALSE"] = 3] = "IS_NOT_FALSE";
    BoolTestType[BoolTestType["IS_UNKNOWN"] = 4] = "IS_UNKNOWN";
    BoolTestType[BoolTestType["IS_NOT_UNKNOWN"] = 5] = "IS_NOT_UNKNOWN";
})(BoolTestType || (exports.BoolTestType = BoolTestType = {}));
var CmdType;
(function (CmdType) {
    CmdType[CmdType["CMD_UNKNOWN"] = 0] = "CMD_UNKNOWN";
    CmdType[CmdType["CMD_SELECT"] = 1] = "CMD_SELECT";
    CmdType[CmdType["CMD_UPDATE"] = 2] = "CMD_UPDATE";
    CmdType[CmdType["CMD_INSERT"] = 3] = "CMD_INSERT";
    CmdType[CmdType["CMD_DELETE"] = 4] = "CMD_DELETE";
    CmdType[CmdType["CMD_UTILITY"] = 5] = "CMD_UTILITY";
    CmdType[CmdType["CMD_NOTHING"] = 6] = "CMD_NOTHING";
})(CmdType || (exports.CmdType = CmdType = {}));
var JoinType;
(function (JoinType) {
    JoinType[JoinType["JOIN_INNER"] = 0] = "JOIN_INNER";
    JoinType[JoinType["JOIN_LEFT"] = 1] = "JOIN_LEFT";
    JoinType[JoinType["JOIN_FULL"] = 2] = "JOIN_FULL";
    JoinType[JoinType["JOIN_RIGHT"] = 3] = "JOIN_RIGHT";
    JoinType[JoinType["JOIN_SEMI"] = 4] = "JOIN_SEMI";
    JoinType[JoinType["JOIN_ANTI"] = 5] = "JOIN_ANTI";
    JoinType[JoinType["JOIN_UNIQUE_OUTER"] = 6] = "JOIN_UNIQUE_OUTER";
    JoinType[JoinType["JOIN_UNIQUE_INNER"] = 7] = "JOIN_UNIQUE_INNER";
})(JoinType || (exports.JoinType = JoinType = {}));
var AggStrategy;
(function (AggStrategy) {
    AggStrategy[AggStrategy["AGG_PLAIN"] = 0] = "AGG_PLAIN";
    AggStrategy[AggStrategy["AGG_SORTED"] = 1] = "AGG_SORTED";
    AggStrategy[AggStrategy["AGG_HASHED"] = 2] = "AGG_HASHED";
    AggStrategy[AggStrategy["AGG_MIXED"] = 3] = "AGG_MIXED";
})(AggStrategy || (exports.AggStrategy = AggStrategy = {}));
var AggSplit;
(function (AggSplit) {
    AggSplit[AggSplit["AGGSPLIT_SIMPLE"] = 0] = "AGGSPLIT_SIMPLE";
    AggSplit[AggSplit["AGGSPLIT_INITIAL_SERIAL"] = 1] = "AGGSPLIT_INITIAL_SERIAL";
    AggSplit[AggSplit["AGGSPLIT_FINAL_DESERIAL"] = 2] = "AGGSPLIT_FINAL_DESERIAL";
})(AggSplit || (exports.AggSplit = AggSplit = {}));
var SetOpCmd;
(function (SetOpCmd) {
    SetOpCmd[SetOpCmd["SETOPCMD_INTERSECT"] = 0] = "SETOPCMD_INTERSECT";
    SetOpCmd[SetOpCmd["SETOPCMD_INTERSECT_ALL"] = 1] = "SETOPCMD_INTERSECT_ALL";
    SetOpCmd[SetOpCmd["SETOPCMD_EXCEPT"] = 2] = "SETOPCMD_EXCEPT";
    SetOpCmd[SetOpCmd["SETOPCMD_EXCEPT_ALL"] = 3] = "SETOPCMD_EXCEPT_ALL";
})(SetOpCmd || (exports.SetOpCmd = SetOpCmd = {}));
var SetOpStrategy;
(function (SetOpStrategy) {
    SetOpStrategy[SetOpStrategy["SETOP_SORTED"] = 0] = "SETOP_SORTED";
    SetOpStrategy[SetOpStrategy["SETOP_HASHED"] = 1] = "SETOP_HASHED";
})(SetOpStrategy || (exports.SetOpStrategy = SetOpStrategy = {}));
var OnConflictAction;
(function (OnConflictAction) {
    OnConflictAction[OnConflictAction["ONCONFLICT_NONE"] = 0] = "ONCONFLICT_NONE";
    OnConflictAction[OnConflictAction["ONCONFLICT_NOTHING"] = 1] = "ONCONFLICT_NOTHING";
    OnConflictAction[OnConflictAction["ONCONFLICT_UPDATE"] = 2] = "ONCONFLICT_UPDATE";
})(OnConflictAction || (exports.OnConflictAction = OnConflictAction = {}));
var LimitOption;
(function (LimitOption) {
    LimitOption[LimitOption["LIMIT_OPTION_DEFAULT"] = 0] = "LIMIT_OPTION_DEFAULT";
    LimitOption[LimitOption["LIMIT_OPTION_COUNT"] = 1] = "LIMIT_OPTION_COUNT";
    LimitOption[LimitOption["LIMIT_OPTION_WITH_TIES"] = 2] = "LIMIT_OPTION_WITH_TIES";
})(LimitOption || (exports.LimitOption = LimitOption = {}));
var LockClauseStrength;
(function (LockClauseStrength) {
    LockClauseStrength[LockClauseStrength["LCS_NONE"] = 0] = "LCS_NONE";
    LockClauseStrength[LockClauseStrength["LCS_FORKEYSHARE"] = 1] = "LCS_FORKEYSHARE";
    LockClauseStrength[LockClauseStrength["LCS_FORSHARE"] = 2] = "LCS_FORSHARE";
    LockClauseStrength[LockClauseStrength["LCS_FORNOKEYUPDATE"] = 3] = "LCS_FORNOKEYUPDATE";
    LockClauseStrength[LockClauseStrength["LCS_FORUPDATE"] = 4] = "LCS_FORUPDATE";
})(LockClauseStrength || (exports.LockClauseStrength = LockClauseStrength = {}));
var LockWaitPolicy;
(function (LockWaitPolicy) {
    LockWaitPolicy[LockWaitPolicy["LockWaitBlock"] = 0] = "LockWaitBlock";
    LockWaitPolicy[LockWaitPolicy["LockWaitSkip"] = 1] = "LockWaitSkip";
    LockWaitPolicy[LockWaitPolicy["LockWaitError"] = 2] = "LockWaitError";
})(LockWaitPolicy || (exports.LockWaitPolicy = LockWaitPolicy = {}));
var LockTupleMode;
(function (LockTupleMode) {
    LockTupleMode[LockTupleMode["LockTupleKeyShare"] = 0] = "LockTupleKeyShare";
    LockTupleMode[LockTupleMode["LockTupleShare"] = 1] = "LockTupleShare";
    LockTupleMode[LockTupleMode["LockTupleNoKeyExclusive"] = 2] = "LockTupleNoKeyExclusive";
    LockTupleMode[LockTupleMode["LockTupleExclusive"] = 3] = "LockTupleExclusive";
})(LockTupleMode || (exports.LockTupleMode = LockTupleMode = {}));
var KeywordKind;
(function (KeywordKind) {
    KeywordKind[KeywordKind["NO_KEYWORD"] = 0] = "NO_KEYWORD";
    KeywordKind[KeywordKind["UNRESERVED_KEYWORD"] = 1] = "UNRESERVED_KEYWORD";
    KeywordKind[KeywordKind["COL_NAME_KEYWORD"] = 2] = "COL_NAME_KEYWORD";
    KeywordKind[KeywordKind["TYPE_FUNC_NAME_KEYWORD"] = 3] = "TYPE_FUNC_NAME_KEYWORD";
    KeywordKind[KeywordKind["RESERVED_KEYWORD"] = 4] = "RESERVED_KEYWORD";
})(KeywordKind || (exports.KeywordKind = KeywordKind = {}));
var Token;
(function (Token) {
    Token[Token["NUL"] = 0] = "NUL";
    Token[Token["ASCII_37"] = 37] = "ASCII_37";
    Token[Token["ASCII_40"] = 40] = "ASCII_40";
    Token[Token["ASCII_41"] = 41] = "ASCII_41";
    Token[Token["ASCII_42"] = 42] = "ASCII_42";
    Token[Token["ASCII_43"] = 43] = "ASCII_43";
    Token[Token["ASCII_44"] = 44] = "ASCII_44";
    Token[Token["ASCII_45"] = 45] = "ASCII_45";
    Token[Token["ASCII_46"] = 46] = "ASCII_46";
    Token[Token["ASCII_47"] = 47] = "ASCII_47";
    Token[Token["ASCII_58"] = 58] = "ASCII_58";
    Token[Token["ASCII_59"] = 59] = "ASCII_59";
    Token[Token["ASCII_60"] = 60] = "ASCII_60";
    Token[Token["ASCII_61"] = 61] = "ASCII_61";
    Token[Token["ASCII_62"] = 62] = "ASCII_62";
    Token[Token["ASCII_63"] = 63] = "ASCII_63";
    Token[Token["ASCII_91"] = 91] = "ASCII_91";
    Token[Token["ASCII_92"] = 92] = "ASCII_92";
    Token[Token["ASCII_93"] = 93] = "ASCII_93";
    Token[Token["ASCII_94"] = 94] = "ASCII_94";
    Token[Token["IDENT"] = 258] = "IDENT";
    Token[Token["UIDENT"] = 259] = "UIDENT";
    Token[Token["FCONST"] = 260] = "FCONST";
    Token[Token["SCONST"] = 261] = "SCONST";
    Token[Token["USCONST"] = 262] = "USCONST";
    Token[Token["BCONST"] = 263] = "BCONST";
    Token[Token["XCONST"] = 264] = "XCONST";
    Token[Token["Op"] = 265] = "Op";
    Token[Token["ICONST"] = 266] = "ICONST";
    Token[Token["PARAM"] = 267] = "PARAM";
    Token[Token["TYPECAST"] = 268] = "TYPECAST";
    Token[Token["DOT_DOT"] = 269] = "DOT_DOT";
    Token[Token["COLON_EQUALS"] = 270] = "COLON_EQUALS";
    Token[Token["EQUALS_GREATER"] = 271] = "EQUALS_GREATER";
    Token[Token["LESS_EQUALS"] = 272] = "LESS_EQUALS";
    Token[Token["GREATER_EQUALS"] = 273] = "GREATER_EQUALS";
    Token[Token["NOT_EQUALS"] = 274] = "NOT_EQUALS";
    Token[Token["SQL_COMMENT"] = 275] = "SQL_COMMENT";
    Token[Token["C_COMMENT"] = 276] = "C_COMMENT";
    Token[Token["ABORT_P"] = 277] = "ABORT_P";
    Token[Token["ABSOLUTE_P"] = 278] = "ABSOLUTE_P";
    Token[Token["ACCESS"] = 279] = "ACCESS";
    Token[Token["ACTION"] = 280] = "ACTION";
    Token[Token["ADD_P"] = 281] = "ADD_P";
    Token[Token["ADMIN"] = 282] = "ADMIN";
    Token[Token["AFTER"] = 283] = "AFTER";
    Token[Token["AGGREGATE"] = 284] = "AGGREGATE";
    Token[Token["ALL"] = 285] = "ALL";
    Token[Token["ALSO"] = 286] = "ALSO";
    Token[Token["ALTER"] = 287] = "ALTER";
    Token[Token["ALWAYS"] = 288] = "ALWAYS";
    Token[Token["ANALYSE"] = 289] = "ANALYSE";
    Token[Token["ANALYZE"] = 290] = "ANALYZE";
    Token[Token["AND"] = 291] = "AND";
    Token[Token["ANY"] = 292] = "ANY";
    Token[Token["ARRAY"] = 293] = "ARRAY";
    Token[Token["AS"] = 294] = "AS";
    Token[Token["ASC"] = 295] = "ASC";
    Token[Token["ASSERTION"] = 296] = "ASSERTION";
    Token[Token["ASSIGNMENT"] = 297] = "ASSIGNMENT";
    Token[Token["ASYMMETRIC"] = 298] = "ASYMMETRIC";
    Token[Token["AT"] = 299] = "AT";
    Token[Token["ATTACH"] = 300] = "ATTACH";
    Token[Token["ATTRIBUTE"] = 301] = "ATTRIBUTE";
    Token[Token["AUTHORIZATION"] = 302] = "AUTHORIZATION";
    Token[Token["BACKWARD"] = 303] = "BACKWARD";
    Token[Token["BEFORE"] = 304] = "BEFORE";
    Token[Token["BEGIN_P"] = 305] = "BEGIN_P";
    Token[Token["BETWEEN"] = 306] = "BETWEEN";
    Token[Token["BIGINT"] = 307] = "BIGINT";
    Token[Token["BINARY"] = 308] = "BINARY";
    Token[Token["BIT"] = 309] = "BIT";
    Token[Token["BOOLEAN_P"] = 310] = "BOOLEAN_P";
    Token[Token["BOTH"] = 311] = "BOTH";
    Token[Token["BY"] = 312] = "BY";
    Token[Token["CACHE"] = 313] = "CACHE";
    Token[Token["CALL"] = 314] = "CALL";
    Token[Token["CALLED"] = 315] = "CALLED";
    Token[Token["CASCADE"] = 316] = "CASCADE";
    Token[Token["CASCADED"] = 317] = "CASCADED";
    Token[Token["CASE"] = 318] = "CASE";
    Token[Token["CAST"] = 319] = "CAST";
    Token[Token["CATALOG_P"] = 320] = "CATALOG_P";
    Token[Token["CHAIN"] = 321] = "CHAIN";
    Token[Token["CHAR_P"] = 322] = "CHAR_P";
    Token[Token["CHARACTER"] = 323] = "CHARACTER";
    Token[Token["CHARACTERISTICS"] = 324] = "CHARACTERISTICS";
    Token[Token["CHECK"] = 325] = "CHECK";
    Token[Token["CHECKPOINT"] = 326] = "CHECKPOINT";
    Token[Token["CLASS"] = 327] = "CLASS";
    Token[Token["CLOSE"] = 328] = "CLOSE";
    Token[Token["CLUSTER"] = 329] = "CLUSTER";
    Token[Token["COALESCE"] = 330] = "COALESCE";
    Token[Token["COLLATE"] = 331] = "COLLATE";
    Token[Token["COLLATION"] = 332] = "COLLATION";
    Token[Token["COLUMN"] = 333] = "COLUMN";
    Token[Token["COLUMNS"] = 334] = "COLUMNS";
    Token[Token["COMMENT"] = 335] = "COMMENT";
    Token[Token["COMMENTS"] = 336] = "COMMENTS";
    Token[Token["COMMIT"] = 337] = "COMMIT";
    Token[Token["COMMITTED"] = 338] = "COMMITTED";
    Token[Token["CONCURRENTLY"] = 339] = "CONCURRENTLY";
    Token[Token["CONFIGURATION"] = 340] = "CONFIGURATION";
    Token[Token["CONFLICT"] = 341] = "CONFLICT";
    Token[Token["CONNECTION"] = 342] = "CONNECTION";
    Token[Token["CONSTRAINT"] = 343] = "CONSTRAINT";
    Token[Token["CONSTRAINTS"] = 344] = "CONSTRAINTS";
    Token[Token["CONTENT_P"] = 345] = "CONTENT_P";
    Token[Token["CONTINUE_P"] = 346] = "CONTINUE_P";
    Token[Token["CONVERSION_P"] = 347] = "CONVERSION_P";
    Token[Token["COPY"] = 348] = "COPY";
    Token[Token["COST"] = 349] = "COST";
    Token[Token["CREATE"] = 350] = "CREATE";
    Token[Token["CROSS"] = 351] = "CROSS";
    Token[Token["CSV"] = 352] = "CSV";
    Token[Token["CUBE"] = 353] = "CUBE";
    Token[Token["CURRENT_P"] = 354] = "CURRENT_P";
    Token[Token["CURRENT_CATALOG"] = 355] = "CURRENT_CATALOG";
    Token[Token["CURRENT_DATE"] = 356] = "CURRENT_DATE";
    Token[Token["CURRENT_ROLE"] = 357] = "CURRENT_ROLE";
    Token[Token["CURRENT_SCHEMA"] = 358] = "CURRENT_SCHEMA";
    Token[Token["CURRENT_TIME"] = 359] = "CURRENT_TIME";
    Token[Token["CURRENT_TIMESTAMP"] = 360] = "CURRENT_TIMESTAMP";
    Token[Token["CURRENT_USER"] = 361] = "CURRENT_USER";
    Token[Token["CURSOR"] = 362] = "CURSOR";
    Token[Token["CYCLE"] = 363] = "CYCLE";
    Token[Token["DATA_P"] = 364] = "DATA_P";
    Token[Token["DATABASE"] = 365] = "DATABASE";
    Token[Token["DAY_P"] = 366] = "DAY_P";
    Token[Token["DEALLOCATE"] = 367] = "DEALLOCATE";
    Token[Token["DEC"] = 368] = "DEC";
    Token[Token["DECIMAL_P"] = 369] = "DECIMAL_P";
    Token[Token["DECLARE"] = 370] = "DECLARE";
    Token[Token["DEFAULT"] = 371] = "DEFAULT";
    Token[Token["DEFAULTS"] = 372] = "DEFAULTS";
    Token[Token["DEFERRABLE"] = 373] = "DEFERRABLE";
    Token[Token["DEFERRED"] = 374] = "DEFERRED";
    Token[Token["DEFINER"] = 375] = "DEFINER";
    Token[Token["DELETE_P"] = 376] = "DELETE_P";
    Token[Token["DELIMITER"] = 377] = "DELIMITER";
    Token[Token["DELIMITERS"] = 378] = "DELIMITERS";
    Token[Token["DEPENDS"] = 379] = "DEPENDS";
    Token[Token["DESC"] = 380] = "DESC";
    Token[Token["DETACH"] = 381] = "DETACH";
    Token[Token["DICTIONARY"] = 382] = "DICTIONARY";
    Token[Token["DISABLE_P"] = 383] = "DISABLE_P";
    Token[Token["DISCARD"] = 384] = "DISCARD";
    Token[Token["DISTINCT"] = 385] = "DISTINCT";
    Token[Token["DO"] = 386] = "DO";
    Token[Token["DOCUMENT_P"] = 387] = "DOCUMENT_P";
    Token[Token["DOMAIN_P"] = 388] = "DOMAIN_P";
    Token[Token["DOUBLE_P"] = 389] = "DOUBLE_P";
    Token[Token["DROP"] = 390] = "DROP";
    Token[Token["EACH"] = 391] = "EACH";
    Token[Token["ELSE"] = 392] = "ELSE";
    Token[Token["ENABLE_P"] = 393] = "ENABLE_P";
    Token[Token["ENCODING"] = 394] = "ENCODING";
    Token[Token["ENCRYPTED"] = 395] = "ENCRYPTED";
    Token[Token["END_P"] = 396] = "END_P";
    Token[Token["ENUM_P"] = 397] = "ENUM_P";
    Token[Token["ESCAPE"] = 398] = "ESCAPE";
    Token[Token["EVENT"] = 399] = "EVENT";
    Token[Token["EXCEPT"] = 400] = "EXCEPT";
    Token[Token["EXCLUDE"] = 401] = "EXCLUDE";
    Token[Token["EXCLUDING"] = 402] = "EXCLUDING";
    Token[Token["EXCLUSIVE"] = 403] = "EXCLUSIVE";
    Token[Token["EXECUTE"] = 404] = "EXECUTE";
    Token[Token["EXISTS"] = 405] = "EXISTS";
    Token[Token["EXPLAIN"] = 406] = "EXPLAIN";
    Token[Token["EXPRESSION"] = 407] = "EXPRESSION";
    Token[Token["EXTENSION"] = 408] = "EXTENSION";
    Token[Token["EXTERNAL"] = 409] = "EXTERNAL";
    Token[Token["EXTRACT"] = 410] = "EXTRACT";
    Token[Token["FALSE_P"] = 411] = "FALSE_P";
    Token[Token["FAMILY"] = 412] = "FAMILY";
    Token[Token["FETCH"] = 413] = "FETCH";
    Token[Token["FILTER"] = 414] = "FILTER";
    Token[Token["FIRST_P"] = 415] = "FIRST_P";
    Token[Token["FLOAT_P"] = 416] = "FLOAT_P";
    Token[Token["FOLLOWING"] = 417] = "FOLLOWING";
    Token[Token["FOR"] = 418] = "FOR";
    Token[Token["FORCE"] = 419] = "FORCE";
    Token[Token["FOREIGN"] = 420] = "FOREIGN";
    Token[Token["FORWARD"] = 421] = "FORWARD";
    Token[Token["FREEZE"] = 422] = "FREEZE";
    Token[Token["FROM"] = 423] = "FROM";
    Token[Token["FULL"] = 424] = "FULL";
    Token[Token["FUNCTION"] = 425] = "FUNCTION";
    Token[Token["FUNCTIONS"] = 426] = "FUNCTIONS";
    Token[Token["GENERATED"] = 427] = "GENERATED";
    Token[Token["GLOBAL"] = 428] = "GLOBAL";
    Token[Token["GRANT"] = 429] = "GRANT";
    Token[Token["GRANTED"] = 430] = "GRANTED";
    Token[Token["GREATEST"] = 431] = "GREATEST";
    Token[Token["GROUP_P"] = 432] = "GROUP_P";
    Token[Token["GROUPING"] = 433] = "GROUPING";
    Token[Token["GROUPS"] = 434] = "GROUPS";
    Token[Token["HANDLER"] = 435] = "HANDLER";
    Token[Token["HAVING"] = 436] = "HAVING";
    Token[Token["HEADER_P"] = 437] = "HEADER_P";
    Token[Token["HOLD"] = 438] = "HOLD";
    Token[Token["HOUR_P"] = 439] = "HOUR_P";
    Token[Token["IDENTITY_P"] = 440] = "IDENTITY_P";
    Token[Token["IF_P"] = 441] = "IF_P";
    Token[Token["ILIKE"] = 442] = "ILIKE";
    Token[Token["IMMEDIATE"] = 443] = "IMMEDIATE";
    Token[Token["IMMUTABLE"] = 444] = "IMMUTABLE";
    Token[Token["IMPLICIT_P"] = 445] = "IMPLICIT_P";
    Token[Token["IMPORT_P"] = 446] = "IMPORT_P";
    Token[Token["IN_P"] = 447] = "IN_P";
    Token[Token["INCLUDE"] = 448] = "INCLUDE";
    Token[Token["INCLUDING"] = 449] = "INCLUDING";
    Token[Token["INCREMENT"] = 450] = "INCREMENT";
    Token[Token["INDEX"] = 451] = "INDEX";
    Token[Token["INDEXES"] = 452] = "INDEXES";
    Token[Token["INHERIT"] = 453] = "INHERIT";
    Token[Token["INHERITS"] = 454] = "INHERITS";
    Token[Token["INITIALLY"] = 455] = "INITIALLY";
    Token[Token["INLINE_P"] = 456] = "INLINE_P";
    Token[Token["INNER_P"] = 457] = "INNER_P";
    Token[Token["INOUT"] = 458] = "INOUT";
    Token[Token["INPUT_P"] = 459] = "INPUT_P";
    Token[Token["INSENSITIVE"] = 460] = "INSENSITIVE";
    Token[Token["INSERT"] = 461] = "INSERT";
    Token[Token["INSTEAD"] = 462] = "INSTEAD";
    Token[Token["INT_P"] = 463] = "INT_P";
    Token[Token["INTEGER"] = 464] = "INTEGER";
    Token[Token["INTERSECT"] = 465] = "INTERSECT";
    Token[Token["INTERVAL"] = 466] = "INTERVAL";
    Token[Token["INTO"] = 467] = "INTO";
    Token[Token["INVOKER"] = 468] = "INVOKER";
    Token[Token["IS"] = 469] = "IS";
    Token[Token["ISNULL"] = 470] = "ISNULL";
    Token[Token["ISOLATION"] = 471] = "ISOLATION";
    Token[Token["JOIN"] = 472] = "JOIN";
    Token[Token["KEY"] = 473] = "KEY";
    Token[Token["LABEL"] = 474] = "LABEL";
    Token[Token["LANGUAGE"] = 475] = "LANGUAGE";
    Token[Token["LARGE_P"] = 476] = "LARGE_P";
    Token[Token["LAST_P"] = 477] = "LAST_P";
    Token[Token["LATERAL_P"] = 478] = "LATERAL_P";
    Token[Token["LEADING"] = 479] = "LEADING";
    Token[Token["LEAKPROOF"] = 480] = "LEAKPROOF";
    Token[Token["LEAST"] = 481] = "LEAST";
    Token[Token["LEFT"] = 482] = "LEFT";
    Token[Token["LEVEL"] = 483] = "LEVEL";
    Token[Token["LIKE"] = 484] = "LIKE";
    Token[Token["LIMIT"] = 485] = "LIMIT";
    Token[Token["LISTEN"] = 486] = "LISTEN";
    Token[Token["LOAD"] = 487] = "LOAD";
    Token[Token["LOCAL"] = 488] = "LOCAL";
    Token[Token["LOCALTIME"] = 489] = "LOCALTIME";
    Token[Token["LOCALTIMESTAMP"] = 490] = "LOCALTIMESTAMP";
    Token[Token["LOCATION"] = 491] = "LOCATION";
    Token[Token["LOCK_P"] = 492] = "LOCK_P";
    Token[Token["LOCKED"] = 493] = "LOCKED";
    Token[Token["LOGGED"] = 494] = "LOGGED";
    Token[Token["MAPPING"] = 495] = "MAPPING";
    Token[Token["MATCH"] = 496] = "MATCH";
    Token[Token["MATERIALIZED"] = 497] = "MATERIALIZED";
    Token[Token["MAXVALUE"] = 498] = "MAXVALUE";
    Token[Token["METHOD"] = 499] = "METHOD";
    Token[Token["MINUTE_P"] = 500] = "MINUTE_P";
    Token[Token["MINVALUE"] = 501] = "MINVALUE";
    Token[Token["MODE"] = 502] = "MODE";
    Token[Token["MONTH_P"] = 503] = "MONTH_P";
    Token[Token["MOVE"] = 504] = "MOVE";
    Token[Token["NAME_P"] = 505] = "NAME_P";
    Token[Token["NAMES"] = 506] = "NAMES";
    Token[Token["NATIONAL"] = 507] = "NATIONAL";
    Token[Token["NATURAL"] = 508] = "NATURAL";
    Token[Token["NCHAR"] = 509] = "NCHAR";
    Token[Token["NEW"] = 510] = "NEW";
    Token[Token["NEXT"] = 511] = "NEXT";
    Token[Token["NFC"] = 512] = "NFC";
    Token[Token["NFD"] = 513] = "NFD";
    Token[Token["NFKC"] = 514] = "NFKC";
    Token[Token["NFKD"] = 515] = "NFKD";
    Token[Token["NO"] = 516] = "NO";
    Token[Token["NONE"] = 517] = "NONE";
    Token[Token["NORMALIZE"] = 518] = "NORMALIZE";
    Token[Token["NORMALIZED"] = 519] = "NORMALIZED";
    Token[Token["NOT"] = 520] = "NOT";
    Token[Token["NOTHING"] = 521] = "NOTHING";
    Token[Token["NOTIFY"] = 522] = "NOTIFY";
    Token[Token["NOTNULL"] = 523] = "NOTNULL";
    Token[Token["NOWAIT"] = 524] = "NOWAIT";
    Token[Token["NULL_P"] = 525] = "NULL_P";
    Token[Token["NULLIF"] = 526] = "NULLIF";
    Token[Token["NULLS_P"] = 527] = "NULLS_P";
    Token[Token["NUMERIC"] = 528] = "NUMERIC";
    Token[Token["OBJECT_P"] = 529] = "OBJECT_P";
    Token[Token["OF"] = 530] = "OF";
    Token[Token["OFF"] = 531] = "OFF";
    Token[Token["OFFSET"] = 532] = "OFFSET";
    Token[Token["OIDS"] = 533] = "OIDS";
    Token[Token["OLD"] = 534] = "OLD";
    Token[Token["ON"] = 535] = "ON";
    Token[Token["ONLY"] = 536] = "ONLY";
    Token[Token["OPERATOR"] = 537] = "OPERATOR";
    Token[Token["OPTION"] = 538] = "OPTION";
    Token[Token["OPTIONS"] = 539] = "OPTIONS";
    Token[Token["OR"] = 540] = "OR";
    Token[Token["ORDER"] = 541] = "ORDER";
    Token[Token["ORDINALITY"] = 542] = "ORDINALITY";
    Token[Token["OTHERS"] = 543] = "OTHERS";
    Token[Token["OUT_P"] = 544] = "OUT_P";
    Token[Token["OUTER_P"] = 545] = "OUTER_P";
    Token[Token["OVER"] = 546] = "OVER";
    Token[Token["OVERLAPS"] = 547] = "OVERLAPS";
    Token[Token["OVERLAY"] = 548] = "OVERLAY";
    Token[Token["OVERRIDING"] = 549] = "OVERRIDING";
    Token[Token["OWNED"] = 550] = "OWNED";
    Token[Token["OWNER"] = 551] = "OWNER";
    Token[Token["PARALLEL"] = 552] = "PARALLEL";
    Token[Token["PARSER"] = 553] = "PARSER";
    Token[Token["PARTIAL"] = 554] = "PARTIAL";
    Token[Token["PARTITION"] = 555] = "PARTITION";
    Token[Token["PASSING"] = 556] = "PASSING";
    Token[Token["PASSWORD"] = 557] = "PASSWORD";
    Token[Token["PLACING"] = 558] = "PLACING";
    Token[Token["PLANS"] = 559] = "PLANS";
    Token[Token["POLICY"] = 560] = "POLICY";
    Token[Token["POSITION"] = 561] = "POSITION";
    Token[Token["PRECEDING"] = 562] = "PRECEDING";
    Token[Token["PRECISION"] = 563] = "PRECISION";
    Token[Token["PRESERVE"] = 564] = "PRESERVE";
    Token[Token["PREPARE"] = 565] = "PREPARE";
    Token[Token["PREPARED"] = 566] = "PREPARED";
    Token[Token["PRIMARY"] = 567] = "PRIMARY";
    Token[Token["PRIOR"] = 568] = "PRIOR";
    Token[Token["PRIVILEGES"] = 569] = "PRIVILEGES";
    Token[Token["PROCEDURAL"] = 570] = "PROCEDURAL";
    Token[Token["PROCEDURE"] = 571] = "PROCEDURE";
    Token[Token["PROCEDURES"] = 572] = "PROCEDURES";
    Token[Token["PROGRAM"] = 573] = "PROGRAM";
    Token[Token["PUBLICATION"] = 574] = "PUBLICATION";
    Token[Token["QUOTE"] = 575] = "QUOTE";
    Token[Token["RANGE"] = 576] = "RANGE";
    Token[Token["READ"] = 577] = "READ";
    Token[Token["REAL"] = 578] = "REAL";
    Token[Token["REASSIGN"] = 579] = "REASSIGN";
    Token[Token["RECHECK"] = 580] = "RECHECK";
    Token[Token["RECURSIVE"] = 581] = "RECURSIVE";
    Token[Token["REF_P"] = 582] = "REF_P";
    Token[Token["REFERENCES"] = 583] = "REFERENCES";
    Token[Token["REFERENCING"] = 584] = "REFERENCING";
    Token[Token["REFRESH"] = 585] = "REFRESH";
    Token[Token["REINDEX"] = 586] = "REINDEX";
    Token[Token["RELATIVE_P"] = 587] = "RELATIVE_P";
    Token[Token["RELEASE"] = 588] = "RELEASE";
    Token[Token["RENAME"] = 589] = "RENAME";
    Token[Token["REPEATABLE"] = 590] = "REPEATABLE";
    Token[Token["REPLACE"] = 591] = "REPLACE";
    Token[Token["REPLICA"] = 592] = "REPLICA";
    Token[Token["RESET"] = 593] = "RESET";
    Token[Token["RESTART"] = 594] = "RESTART";
    Token[Token["RESTRICT"] = 595] = "RESTRICT";
    Token[Token["RETURNING"] = 596] = "RETURNING";
    Token[Token["RETURNS"] = 597] = "RETURNS";
    Token[Token["REVOKE"] = 598] = "REVOKE";
    Token[Token["RIGHT"] = 599] = "RIGHT";
    Token[Token["ROLE"] = 600] = "ROLE";
    Token[Token["ROLLBACK"] = 601] = "ROLLBACK";
    Token[Token["ROLLUP"] = 602] = "ROLLUP";
    Token[Token["ROUTINE"] = 603] = "ROUTINE";
    Token[Token["ROUTINES"] = 604] = "ROUTINES";
    Token[Token["ROW"] = 605] = "ROW";
    Token[Token["ROWS"] = 606] = "ROWS";
    Token[Token["RULE"] = 607] = "RULE";
    Token[Token["SAVEPOINT"] = 608] = "SAVEPOINT";
    Token[Token["SCHEMA"] = 609] = "SCHEMA";
    Token[Token["SCHEMAS"] = 610] = "SCHEMAS";
    Token[Token["SCROLL"] = 611] = "SCROLL";
    Token[Token["SEARCH"] = 612] = "SEARCH";
    Token[Token["SECOND_P"] = 613] = "SECOND_P";
    Token[Token["SECURITY"] = 614] = "SECURITY";
    Token[Token["SELECT"] = 615] = "SELECT";
    Token[Token["SEQUENCE"] = 616] = "SEQUENCE";
    Token[Token["SEQUENCES"] = 617] = "SEQUENCES";
    Token[Token["SERIALIZABLE"] = 618] = "SERIALIZABLE";
    Token[Token["SERVER"] = 619] = "SERVER";
    Token[Token["SESSION"] = 620] = "SESSION";
    Token[Token["SESSION_USER"] = 621] = "SESSION_USER";
    Token[Token["SET"] = 622] = "SET";
    Token[Token["SETS"] = 623] = "SETS";
    Token[Token["SETOF"] = 624] = "SETOF";
    Token[Token["SHARE"] = 625] = "SHARE";
    Token[Token["SHOW"] = 626] = "SHOW";
    Token[Token["SIMILAR"] = 627] = "SIMILAR";
    Token[Token["SIMPLE"] = 628] = "SIMPLE";
    Token[Token["SKIP"] = 629] = "SKIP";
    Token[Token["SMALLINT"] = 630] = "SMALLINT";
    Token[Token["SNAPSHOT"] = 631] = "SNAPSHOT";
    Token[Token["SOME"] = 632] = "SOME";
    Token[Token["SQL_P"] = 633] = "SQL_P";
    Token[Token["STABLE"] = 634] = "STABLE";
    Token[Token["STANDALONE_P"] = 635] = "STANDALONE_P";
    Token[Token["START"] = 636] = "START";
    Token[Token["STATEMENT"] = 637] = "STATEMENT";
    Token[Token["STATISTICS"] = 638] = "STATISTICS";
    Token[Token["STDIN"] = 639] = "STDIN";
    Token[Token["STDOUT"] = 640] = "STDOUT";
    Token[Token["STORAGE"] = 641] = "STORAGE";
    Token[Token["STORED"] = 642] = "STORED";
    Token[Token["STRICT_P"] = 643] = "STRICT_P";
    Token[Token["STRIP_P"] = 644] = "STRIP_P";
    Token[Token["SUBSCRIPTION"] = 645] = "SUBSCRIPTION";
    Token[Token["SUBSTRING"] = 646] = "SUBSTRING";
    Token[Token["SUPPORT"] = 647] = "SUPPORT";
    Token[Token["SYMMETRIC"] = 648] = "SYMMETRIC";
    Token[Token["SYSID"] = 649] = "SYSID";
    Token[Token["SYSTEM_P"] = 650] = "SYSTEM_P";
    Token[Token["TABLE"] = 651] = "TABLE";
    Token[Token["TABLES"] = 652] = "TABLES";
    Token[Token["TABLESAMPLE"] = 653] = "TABLESAMPLE";
    Token[Token["TABLESPACE"] = 654] = "TABLESPACE";
    Token[Token["TEMP"] = 655] = "TEMP";
    Token[Token["TEMPLATE"] = 656] = "TEMPLATE";
    Token[Token["TEMPORARY"] = 657] = "TEMPORARY";
    Token[Token["TEXT_P"] = 658] = "TEXT_P";
    Token[Token["THEN"] = 659] = "THEN";
    Token[Token["TIES"] = 660] = "TIES";
    Token[Token["TIME"] = 661] = "TIME";
    Token[Token["TIMESTAMP"] = 662] = "TIMESTAMP";
    Token[Token["TO"] = 663] = "TO";
    Token[Token["TRAILING"] = 664] = "TRAILING";
    Token[Token["TRANSACTION"] = 665] = "TRANSACTION";
    Token[Token["TRANSFORM"] = 666] = "TRANSFORM";
    Token[Token["TREAT"] = 667] = "TREAT";
    Token[Token["TRIGGER"] = 668] = "TRIGGER";
    Token[Token["TRIM"] = 669] = "TRIM";
    Token[Token["TRUE_P"] = 670] = "TRUE_P";
    Token[Token["TRUNCATE"] = 671] = "TRUNCATE";
    Token[Token["TRUSTED"] = 672] = "TRUSTED";
    Token[Token["TYPE_P"] = 673] = "TYPE_P";
    Token[Token["TYPES_P"] = 674] = "TYPES_P";
    Token[Token["UESCAPE"] = 675] = "UESCAPE";
    Token[Token["UNBOUNDED"] = 676] = "UNBOUNDED";
    Token[Token["UNCOMMITTED"] = 677] = "UNCOMMITTED";
    Token[Token["UNENCRYPTED"] = 678] = "UNENCRYPTED";
    Token[Token["UNION"] = 679] = "UNION";
    Token[Token["UNIQUE"] = 680] = "UNIQUE";
    Token[Token["UNKNOWN"] = 681] = "UNKNOWN";
    Token[Token["UNLISTEN"] = 682] = "UNLISTEN";
    Token[Token["UNLOGGED"] = 683] = "UNLOGGED";
    Token[Token["UNTIL"] = 684] = "UNTIL";
    Token[Token["UPDATE"] = 685] = "UPDATE";
    Token[Token["USER"] = 686] = "USER";
    Token[Token["USING"] = 687] = "USING";
    Token[Token["VACUUM"] = 688] = "VACUUM";
    Token[Token["VALID"] = 689] = "VALID";
    Token[Token["VALIDATE"] = 690] = "VALIDATE";
    Token[Token["VALIDATOR"] = 691] = "VALIDATOR";
    Token[Token["VALUE_P"] = 692] = "VALUE_P";
    Token[Token["VALUES"] = 693] = "VALUES";
    Token[Token["VARCHAR"] = 694] = "VARCHAR";
    Token[Token["VARIADIC"] = 695] = "VARIADIC";
    Token[Token["VARYING"] = 696] = "VARYING";
    Token[Token["VERBOSE"] = 697] = "VERBOSE";
    Token[Token["VERSION_P"] = 698] = "VERSION_P";
    Token[Token["VIEW"] = 699] = "VIEW";
    Token[Token["VIEWS"] = 700] = "VIEWS";
    Token[Token["VOLATILE"] = 701] = "VOLATILE";
    Token[Token["WHEN"] = 702] = "WHEN";
    Token[Token["WHERE"] = 703] = "WHERE";
    Token[Token["WHITESPACE_P"] = 704] = "WHITESPACE_P";
    Token[Token["WINDOW"] = 705] = "WINDOW";
    Token[Token["WITH"] = 706] = "WITH";
    Token[Token["WITHIN"] = 707] = "WITHIN";
    Token[Token["WITHOUT"] = 708] = "WITHOUT";
    Token[Token["WORK"] = 709] = "WORK";
    Token[Token["WRAPPER"] = 710] = "WRAPPER";
    Token[Token["WRITE"] = 711] = "WRITE";
    Token[Token["XML_P"] = 712] = "XML_P";
    Token[Token["XMLATTRIBUTES"] = 713] = "XMLATTRIBUTES";
    Token[Token["XMLCONCAT"] = 714] = "XMLCONCAT";
    Token[Token["XMLELEMENT"] = 715] = "XMLELEMENT";
    Token[Token["XMLEXISTS"] = 716] = "XMLEXISTS";
    Token[Token["XMLFOREST"] = 717] = "XMLFOREST";
    Token[Token["XMLNAMESPACES"] = 718] = "XMLNAMESPACES";
    Token[Token["XMLPARSE"] = 719] = "XMLPARSE";
    Token[Token["XMLPI"] = 720] = "XMLPI";
    Token[Token["XMLROOT"] = 721] = "XMLROOT";
    Token[Token["XMLSERIALIZE"] = 722] = "XMLSERIALIZE";
    Token[Token["XMLTABLE"] = 723] = "XMLTABLE";
    Token[Token["YEAR_P"] = 724] = "YEAR_P";
    Token[Token["YES_P"] = 725] = "YES_P";
    Token[Token["ZONE"] = 726] = "ZONE";
    Token[Token["NOT_LA"] = 727] = "NOT_LA";
    Token[Token["NULLS_LA"] = 728] = "NULLS_LA";
    Token[Token["WITH_LA"] = 729] = "WITH_LA";
    Token[Token["POSTFIXOP"] = 730] = "POSTFIXOP";
    Token[Token["UMINUS"] = 731] = "UMINUS";
})(Token || (exports.Token = Token = {}));
